File: packages/core/lib/src/models/shape.dart
import 'package:equatable/equatable.dart';
import 'package:flutter/rendering.dart';
import 'package:vio_core/vio_core.dart';

/// Sentinel value for copyWith parameters where we need to distinguish between
/// "not provided" and "explicitly set to null".
///
/// Example: `shape.copyWith(frameId: null)` should clear the frameId, while
/// omitting `frameId` should keep the existing value.
const Object kUnset = Object();

/// Base class for all shape types on the canvas
///
/// This mirrors Penpot's shape model from common/src/app/common/types/shape.cljc
/// All shapes have:
/// - Unique ID
/// - Transform matrix for position/rotation/scale
/// - Selection rectangle (selrect)
/// - Fill and stroke properties
/// - Parent frame reference
abstract class Shape extends Equatable {
  const Shape({
    required this.id,
    required this.name,
    required this.type,
    this.parentId,
    this.frameId,
    this.sortOrder = 0,
    this.transform = Matrix2D.identity,
    this.transformInverse,
    this.selrect,
    this.fills = const [],
    this.strokes = const [],
    this.opacity = 1.0,
    this.hidden = false,
    this.blocked = false,
    this.rotation = 0.0,
    this.constraints,
    this.shadow,
    this.blur,
  });

  /// Unique identifier for this shape
  final String id;

  /// Display name of the shape
  final String name;

  /// Type discriminator (rectangle, ellipse, path, text, frame, etc.)
  final ShapeType type;

  /// Parent shape ID (for groups/frames)
  final String? parentId;

  /// Frame this shape belongs to
  final String? frameId;

  /// Explicit z-order among siblings (lower renders behind higher).
  ///
  /// This is interpreted within the shape's effective container:
  /// - `parentId` when it points to a `GroupShape`
  /// - otherwise `frameId` when it points to a `FrameShape`
  /// - otherwise the root canvas
  final int sortOrder;

  /// Transformation matrix (position, rotation, scale, skew)
  final Matrix2D transform;

  /// Cached inverse of transform matrix
  final Matrix2D? transformInverse;

  /// Selection rectangle in parent coordinates
  final Rect? selrect;

  /// Fill styles
  final List<ShapeFill> fills;

  /// Stroke styles
  final List<ShapeStroke> strokes;

  /// Opacity (0.0 - 1.0)
  final double opacity;

  /// Whether shape is hidden
  final bool hidden;

  /// Whether shape is locked from editing
  final bool blocked;

  /// Rotation in degrees
  final double rotation;

  /// Layout constraints
  final ShapeConstraints? constraints;

  /// Shadow effect
  final ShapeShadow? shadow;

  /// Blur effect
  final ShapeBlur? blur;

  /// Get the bounds of this shape in local coordinates
  Rect get bounds;

  /// X position (abstract - implemented by concrete shape types)
  double get x;

  /// Y position (abstract - implemented by concrete shape types)
  double get y;

  /// Create a copy of this shape moved by the given delta
  Shape moveBy(double dx, double dy);

  /// Create a duplicate of this shape with a new ID and optional position offset
  /// Used for copy/paste and duplicate operations
  Shape duplicate({
    required String newId,
    double offsetX = 0,
    double offsetY = 0,
    String? newName,
  }) {
    final baseDuplicate = copyWith(
      id: newId,
      name: newName ?? '$name Copy',
    );
    if (offsetX != 0 || offsetY != 0) {
      return baseDuplicate.moveBy(offsetX, offsetY);
    }
    return baseDuplicate;
  }

  /// Get the center point in local coordinates
  Offset get center => bounds.center;

  /// Get the width
  double get width => bounds.width;

  /// Get the height
  double get height => bounds.height;

  /// Get position (top-left)
  Offset get position => bounds.topLeft;

  /// Transform a point from local to parent coordinates
  Offset transformPoint(Offset local) {
    final result = transform.transformPoint(local.dx, local.dy);
    return Offset(result.x, result.y);
  }

  /// Transform a point from parent to local coordinates
  Offset inverseTransformPoint(Offset parent) {
    final inverse = transformInverse ?? transform.inverse;
    if (inverse == null) return parent;
    final result = inverse.transformPoint(parent.dx, parent.dy);
    return Offset(result.x, result.y);
  }

  /// Create a copy with updated properties
  Shape copyWith({
    String? id,
    String? name,
    Object? parentId = kUnset,
    Object? frameId = kUnset,
    int? sortOrder,
    Matrix2D? transform,
    Matrix2D? transformInverse,
    Rect? selrect,
    List<ShapeFill>? fills,
    List<ShapeStroke>? strokes,
    double? opacity,
    bool? hidden,
    bool? blocked,
    double? rotation,
    ShapeConstraints? constraints,
    ShapeShadow? shadow,
    ShapeBlur? blur,
  });

  /// Convert shape to JSON for serialization
  Map<String, dynamic> toJson();

  /// Create a shape from JSON data
  static Shape fromJson(Map<String, dynamic> json) {
    final type = ShapeType.values.firstWhere(
      (t) => t.name == json['type'],
      orElse: () => ShapeType.rectangle,
    );

    switch (type) {
      case ShapeType.rectangle:
        return RectangleShape.fromJson(json);
      case ShapeType.ellipse:
        return EllipseShape.fromJson(json);
      case ShapeType.frame:
        return FrameShape.fromJson(json);
      case ShapeType.text:
        return TextShape.fromJson(json);
      case ShapeType.group:
        return GroupShape.fromJson(json);
      case ShapeType.path:
        return PathShape.fromJson(json);
      case ShapeType.image:
        return ImageShape.fromJson(json);
      case ShapeType.svg:
        return SvgShape.fromJson(json);
      case ShapeType.bool:
        return BoolShape.fromJson(json);
    }
  }

  /// Helper to serialize base shape properties
  Map<String, dynamic> baseToJson() => {
        'id': id,
        'name': name,
        'type': type.name,
        if (parentId != null) 'parentId': parentId,
        if (frameId != null) 'frameId': frameId,
        'sortOrder': sortOrder,
        'transform': transform.toJson(),
        if (transformInverse != null)
          'transformInverse': transformInverse!.toJson(),
        if (selrect != null)
          'selrect': {
            'left': selrect!.left,
            'top': selrect!.top,
            'right': selrect!.right,
            'bottom': selrect!.bottom,
          },
        'fills': fills.map((f) => f.toJson()).toList(),
        'strokes': strokes.map((s) => s.toJson()).toList(),
        'opacity': opacity,
        'hidden': hidden,
        'blocked': blocked,
        'rotation': rotation,
        if (constraints != null) 'constraints': constraints!.toJson(),
        if (shadow != null) 'shadow': shadow!.toJson(),
        if (blur != null) 'blur': blur!.toJson(),
      };

  @override
  List<Object?> get props => [
        id,
        name,
        type,
        parentId,
        frameId,
        sortOrder,
        transform,
        selrect,
        fills,
        strokes,
        opacity,
        hidden,
        blocked,
        rotation,
        constraints,
        shadow,
        blur,
      ];
}

/// Shape type discriminator
enum ShapeType {
  rectangle,
  ellipse,
  path,
  text,
  frame,
  group,
  image,
  svg,
  bool,
}

/// Fill style for shapes
class ShapeFill extends Equatable {
  const ShapeFill({
    required this.color,
    this.opacity = 1.0,
    this.hidden = false,
    this.gradient,
    this.fillImage,
  });

  /// Solid color (hex)
  final int color;

  /// Fill opacity
  final double opacity;

  /// Whether this fill is hidden
  final bool hidden;

  /// Gradient fill (optional)
  final ShapeGradient? gradient;

  /// Image fill (optional)
  final ShapeFillImage? fillImage;

  /// Create a copy with updated properties
  ShapeFill copyWith({
    int? color,
    double? opacity,
    bool? hidden,
    ShapeGradient? gradient,
    ShapeFillImage? fillImage,
  }) {
    return ShapeFill(
      color: color ?? this.color,
      opacity: opacity ?? this.opacity,
      hidden: hidden ?? this.hidden,
      gradient: gradient ?? this.gradient,
      fillImage: fillImage ?? this.fillImage,
    );
  }

  Map<String, dynamic> toJson() => {
        'color': color,
        'opacity': opacity,
        'hidden': hidden,
        if (gradient != null) 'gradient': gradient!.toJson(),
        if (fillImage != null) 'fillImage': fillImage!.toJson(),
      };

  factory ShapeFill.fromJson(Map<String, dynamic> json) => ShapeFill(
        color: json['color'] as int,
        opacity: (json['opacity'] as num?)?.toDouble() ?? 1.0,
        hidden: json['hidden'] as bool? ?? false,
        gradient: json['gradient'] != null
            ? ShapeGradient.fromJson(json['gradient'] as Map<String, dynamic>)
            : null,
        fillImage: json['fillImage'] != null
            ? ShapeFillImage.fromJson(json['fillImage'] as Map<String, dynamic>)
            : null,
      );

  @override
  List<Object?> get props => [color, opacity, hidden, gradient, fillImage];
}

/// Stroke style for shapes
class ShapeStroke extends Equatable {
  const ShapeStroke({
    required this.color,
    this.width = 1.0,
    this.opacity = 1.0,
    this.hidden = false,
    this.alignment = StrokeAlignment.center,
    this.cap = StrokeCap.round,
    this.join = StrokeJoin.round,
  });

  /// Stroke color (hex)
  final int color;

  /// Stroke width
  final double width;

  /// Stroke opacity
  final double opacity;

  /// Whether this stroke is hidden
  final bool hidden;

  /// Stroke alignment (inside, center, outside)
  final StrokeAlignment alignment;

  /// Line cap style
  final StrokeCap cap;

  /// Line join style
  final StrokeJoin join;

  /// Create a copy with updated properties
  ShapeStroke copyWith({
    int? color,
    double? width,
    double? opacity,
    bool? hidden,
    StrokeAlignment? alignment,
    StrokeCap? cap,
    StrokeJoin? join,
  }) {
    return ShapeStroke(
      color: color ?? this.color,
      width: width ?? this.width,
      opacity: opacity ?? this.opacity,
      hidden: hidden ?? this.hidden,
      alignment: alignment ?? this.alignment,
      cap: cap ?? this.cap,
      join: join ?? this.join,
    );
  }

  Map<String, dynamic> toJson() => {
        'color': color,
        'width': width,
        'opacity': opacity,
        'hidden': hidden,
        'alignment': alignment.name,
        'cap': cap.name,
        'join': join.name,
      };

  factory ShapeStroke.fromJson(Map<String, dynamic> json) => ShapeStroke(
        color: json['color'] as int,
        width: (json['width'] as num?)?.toDouble() ?? 1.0,
        opacity: (json['opacity'] as num?)?.toDouble() ?? 1.0,
        hidden: json['hidden'] as bool? ?? false,
        alignment: StrokeAlignment.values.firstWhere(
          (e) => e.name == json['alignment'],
          orElse: () => StrokeAlignment.center,
        ),
        cap: StrokeCap.values.firstWhere(
          (e) => e.name == json['cap'],
          orElse: () => StrokeCap.round,
        ),
        join: StrokeJoin.values.firstWhere(
          (e) => e.name == json['join'],
          orElse: () => StrokeJoin.round,
        ),
      );

  @override
  List<Object?> get props =>
      [color, width, opacity, hidden, alignment, cap, join];
}

/// Stroke alignment options
enum StrokeAlignment { inside, center, outside }

/// Stroke cap styles
enum StrokeCap { butt, round, square }

/// Stroke join styles
enum StrokeJoin { miter, round, bevel }

/// Gradient definition
class ShapeGradient extends Equatable {
  const ShapeGradient({
    required this.type,
    required this.stops,
    this.startX = 0.0,
    this.startY = 0.0,
    this.endX = 1.0,
    this.endY = 1.0,
  });

  final GradientType type;
  final List<GradientStop> stops;
  final double startX;
  final double startY;
  final double endX;
  final double endY;

  Map<String, dynamic> toJson() => {
        'type': type.name,
        'stops': stops.map((s) => s.toJson()).toList(),
        'startX': startX,
        'startY': startY,
        'endX': endX,
        'endY': endY,
      };

  factory ShapeGradient.fromJson(Map<String, dynamic> json) => ShapeGradient(
        type: GradientType.values.firstWhere(
          (e) => e.name == json['type'],
          orElse: () => GradientType.linear,
        ),
        stops: (json['stops'] as List)
            .map((s) => GradientStop.fromJson(s as Map<String, dynamic>))
            .toList(),
        startX: (json['startX'] as num?)?.toDouble() ?? 0.0,
        startY: (json['startY'] as num?)?.toDouble() ?? 0.0,
        endX: (json['endX'] as num?)?.toDouble() ?? 1.0,
        endY: (json['endY'] as num?)?.toDouble() ?? 1.0,
      );

  @override
  List<Object?> get props => [type, stops, startX, startY, endX, endY];
}

/// Gradient types
enum GradientType { linear, radial }

/// Gradient color stop
class GradientStop extends Equatable {
  const GradientStop({
    required this.color,
    required this.offset,
    this.opacity = 1.0,
  });

  final int color;
  final double offset;
  final double opacity;

  Map<String, dynamic> toJson() => {
        'color': color,
        'offset': offset,
        'opacity': opacity,
      };

  factory GradientStop.fromJson(Map<String, dynamic> json) => GradientStop(
        color: json['color'] as int,
        offset: (json['offset'] as num).toDouble(),
        opacity: (json['opacity'] as num?)?.toDouble() ?? 1.0,
      );

  @override
  List<Object?> get props => [color, offset, opacity];
}

/// Image fill definition
class ShapeFillImage extends Equatable {
  const ShapeFillImage({required this.id, this.width, this.height, this.mtype});

  final String id;
  final double? width;
  final double? height;
  final String? mtype;

  Map<String, dynamic> toJson() => {
        'id': id,
        if (width != null) 'width': width,
        if (height != null) 'height': height,
        if (mtype != null) 'mtype': mtype,
      };

  factory ShapeFillImage.fromJson(Map<String, dynamic> json) => ShapeFillImage(
        id: json['id'] as String,
        width: (json['width'] as num?)?.toDouble(),
        height: (json['height'] as num?)?.toDouble(),
        mtype: json['mtype'] as String?,
      );

  @override
  List<Object?> get props => [id, width, height, mtype];
}

/// Layout constraints
class ShapeConstraints extends Equatable {
  const ShapeConstraints({
    this.horizontal = ConstraintType.left,
    this.vertical = ConstraintType.top,
  });

  final ConstraintType horizontal;
  final ConstraintType vertical;

  Map<String, dynamic> toJson() => {
        'horizontal': horizontal.name,
        'vertical': vertical.name,
      };

  factory ShapeConstraints.fromJson(Map<String, dynamic> json) =>
      ShapeConstraints(
        horizontal: ConstraintType.values.firstWhere(
          (e) => e.name == json['horizontal'],
          orElse: () => ConstraintType.left,
        ),
        vertical: ConstraintType.values.firstWhere(
          (e) => e.name == json['vertical'],
          orElse: () => ConstraintType.top,
        ),
      );

  @override
  List<Object?> get props => [horizontal, vertical];
}

/// Constraint types
enum ConstraintType {
  left,
  right,
  leftAndRight,
  center,
  scale,
  top,
  bottom,
  topAndBottom,
}

/// Shadow effect
class ShapeShadow extends Equatable {
  const ShapeShadow({
    this.id,
    this.style = ShadowStyle.dropShadow,
    this.color = 0x000000,
    this.opacity = 0.25,
    this.offsetX = 0.0,
    this.offsetY = 4.0,
    this.blur = 8.0,
    this.spread = 0.0,
    this.hidden = false,
  });

  final String? id;
  final ShadowStyle style;
  final int color;
  final double opacity;
  final double offsetX;
  final double offsetY;
  final double blur;
  final double spread;
  final bool hidden;

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        'style': style.name,
        'color': color,
        'opacity': opacity,
        'offsetX': offsetX,
        'offsetY': offsetY,
        'blur': blur,
        'spread': spread,
        'hidden': hidden,
      };

  factory ShapeShadow.fromJson(Map<String, dynamic> json) => ShapeShadow(
        id: json['id'] as String?,
        style: ShadowStyle.values.firstWhere(
          (e) => e.name == json['style'],
          orElse: () => ShadowStyle.dropShadow,
        ),
        color: json['color'] as int? ?? 0x000000,
        opacity: (json['opacity'] as num?)?.toDouble() ?? 0.25,
        offsetX: (json['offsetX'] as num?)?.toDouble() ?? 0.0,
        offsetY: (json['offsetY'] as num?)?.toDouble() ?? 4.0,
        blur: (json['blur'] as num?)?.toDouble() ?? 8.0,
        spread: (json['spread'] as num?)?.toDouble() ?? 0.0,
        hidden: json['hidden'] as bool? ?? false,
      );

  @override
  List<Object?> get props => [
        id,
        style,
        color,
        opacity,
        offsetX,
        offsetY,
        blur,
        spread,
        hidden,
      ];
}

/// Shadow styles
enum ShadowStyle { dropShadow, innerShadow }

/// Blur effect
class ShapeBlur extends Equatable {
  const ShapeBlur({
    this.id,
    this.type = BlurType.layer,
    this.value = 0.0,
    this.hidden = false,
  });

  final String? id;
  final BlurType type;
  final double value;
  final bool hidden;

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        'type': type.name,
        'value': value,
        'hidden': hidden,
      };

  factory ShapeBlur.fromJson(Map<String, dynamic> json) => ShapeBlur(
        id: json['id'] as String?,
        type: BlurType.values.firstWhere(
          (e) => e.name == json['type'],
          orElse: () => BlurType.layer,
        ),
        value: (json['value'] as num?)?.toDouble() ?? 0.0,
        hidden: json['hidden'] as bool? ?? false,
      );

  @override
  List<Object?> get props => [id, type, value, hidden];
}

/// Blur types
enum BlurType { layer, background }

File: apps/client/rust/src/scene_graph/shape.rs
use crate::math::matrix2d::Matrix2D;

/// Mirrors the Dart `ShapeType` enum from packages/core.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ShapeType {
    Rectangle,
    Ellipse,
    Text,
    Frame,
    Group,
    Path,
    Image,
    Svg,
    Bool,
}

/// Text alignment, mirrors Dart's `TextAlign`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TextAlign {
    Left,
    Center,
    Right,
    Justify,
}

/// Boolean operation type.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BoolOp {
    Union,
    Subtract,
    Intersect,
    Exclude,
}

/// Stroke alignment relative to path.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StrokeAlignment {
    Center,
    Inside,
    Outside,
}

/// Gradient type.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GradientType {
    Linear,
    Radial,
}

/// Shadow style.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ShadowStyle {
    Drop,
    Inner,
}

/// Blur type.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BlurType {
    Layer,
    Background,
}

/// Stroke cap.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StrokeCap {
    Butt,
    Round,
    Square,
}

/// Stroke join.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StrokeJoin {
    Miter,
    Round,
    Bevel,
}

/// A gradient stop (offset + color).
#[derive(Debug, Clone, PartialEq)]
pub struct GradientStop {
    pub offset: f64,
    pub color: u32,
}

/// Gradient definition.
#[derive(Debug, Clone, PartialEq)]
pub struct ShapeGradient {
    pub gradient_type: GradientType,
    pub stops: Vec<GradientStop>,
    pub start_x: f64,
    pub start_y: f64,
    pub end_x: f64,
    pub end_y: f64,
}

/// A single fill layer.
#[derive(Debug, Clone, PartialEq)]
pub struct ShapeFill {
    pub color: u32,
    pub opacity: f64,
    pub hidden: bool,
    pub gradient: Option<ShapeGradient>,
}

/// A single stroke layer.
#[derive(Debug, Clone, PartialEq)]
pub struct ShapeStroke {
    pub color: u32,
    pub width: f64,
    pub opacity: f64,
    pub hidden: bool,
    pub alignment: StrokeAlignment,
    pub cap: StrokeCap,
    pub join: StrokeJoin,
}

/// Shadow effect.
#[derive(Debug, Clone, PartialEq)]
pub struct ShapeShadow {
    pub style: ShadowStyle,
    pub color: u32,
    pub opacity: f64,
    pub offset_x: f64,
    pub offset_y: f64,
    pub blur: f64,
    pub spread: f64,
    pub hidden: bool,
}

/// Blur effect.
#[derive(Debug, Clone, PartialEq)]
pub struct ShapeBlur {
    pub blur_type: BlurType,
    pub value: f64,
    pub hidden: bool,
}

/// Type-specific geometric data for each shape variant.
#[derive(Debug, Clone, PartialEq)]
pub enum ShapeGeometry {
    Rectangle {
        width: f64,
        height: f64,
        r1: f64,
        r2: f64,
        r3: f64,
        r4: f64,
    },
    Ellipse {
        width: f64,
        height: f64,
    },
    Text {
        width: f64,
        height: f64,
        text: String,
        font_size: f64,
        font_family: String,
        font_weight: u16,
        line_height: f64,
        letter_spacing_percent: f64,
        text_align: TextAlign,
    },
    Frame {
        width: f64,
        height: f64,
        clip_content: bool,
    },
    Group {
        width: f64,
        height: f64,
    },
    Path {
        width: f64,
        height: f64,
        path_data: String,
        closed: bool,
    },
    Image {
        width: f64,
        height: f64,
        asset_id: String,
    },
    Svg {
        width: f64,
        height: f64,
        svg_content: String,
    },
    Bool {
        width: f64,
        height: f64,
        operation: BoolOp,
    },
}

impl ShapeGeometry {
    /// Get the local-space width and height of the geometry.
    pub fn dimensions(&self) -> (f64, f64) {
        match self {
            ShapeGeometry::Rectangle { width, height, .. } => (*width, *height),
            ShapeGeometry::Ellipse { width, height } => (*width, *height),
            ShapeGeometry::Text { width, height, .. } => (*width, *height),
            ShapeGeometry::Frame { width, height, .. } => (*width, *height),
            ShapeGeometry::Group { width, height } => (*width, *height),
            ShapeGeometry::Path { width, height, .. } => (*width, *height),
            ShapeGeometry::Image { width, height, .. } => (*width, *height),
            ShapeGeometry::Svg { width, height, .. } => (*width, *height),
            ShapeGeometry::Bool { width, height, .. } => (*width, *height),
        }
    }
}

/// A render-optimized shape. Mirrors the Dart `Shape` hierarchy but only
/// carries fields needed for rendering and spatial queries.
#[derive(Debug, Clone, PartialEq)]
pub struct RenderShape {
    pub id: String,
    pub shape_type: ShapeType,
    pub transform: Matrix2D,
    pub parent_id: Option<String>,
    pub frame_id: Option<String>,
    pub sort_order: i32,
    pub opacity: f64,
    pub hidden: bool,
    pub rotation: f64,
    pub fills: Vec<ShapeFill>,
    pub strokes: Vec<ShapeStroke>,
    pub shadow: Option<ShapeShadow>,
    pub blur: Option<ShapeBlur>,
    pub geometry: ShapeGeometry,
}

impl RenderShape {
    /// Get the local-space bounding rect as (x, y, w, h).
    /// Position comes from transform (e, f).
    pub fn local_bounds(&self) -> (f64, f64, f64, f64) {
        let (w, h) = self.geometry.dimensions();
        (0.0, 0.0, w, h)
    }

    /// Get the world-space axis-aligned bounding box.
    pub fn world_aabb(&self) -> crate::math::aabb::Aabb {
        let (_, _, w, h) = self.local_bounds();
        let (bx, by, bw, bh) = self.transform.transform_rect(0.0, 0.0, w, h);
        crate::math::aabb::Aabb::from_xywh(bx, by, bw, bh)
    }

    /// Check if this is a container (Group or Frame).
    pub fn is_container(&self) -> bool {
        matches!(self.shape_type, ShapeType::Group | ShapeType::Frame)
    }

    /// Check if this is a frame with clip_content enabled.
    pub fn clips_content(&self) -> bool {
        matches!(self.geometry, ShapeGeometry::Frame { clip_content: true, .. })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_rect_shape(id: &str, x: f64, y: f64, w: f64, h: f64) -> RenderShape {
        RenderShape {
            id: id.to_string(),
            shape_type: ShapeType::Rectangle,
            transform: Matrix2D::translation(x, y),
            parent_id: None,
            frame_id: None,
            sort_order: 0,
            opacity: 1.0,
            hidden: false,
            rotation: 0.0,
            fills: vec![],
            strokes: vec![],
            shadow: None,
            blur: None,
            geometry: ShapeGeometry::Rectangle {
                width: w,
                height: h,
                r1: 0.0,
                r2: 0.0,
                r3: 0.0,
                r4: 0.0,
            },
        }
    }

    #[test]
    fn local_bounds_of_rectangle() {
        let s = make_rect_shape("r1", 100.0, 200.0, 50.0, 30.0);
        let (x, y, w, h) = s.local_bounds();
        assert_eq!(x, 0.0);
        assert_eq!(y, 0.0);
        assert_eq!(w, 50.0);
        assert_eq!(h, 30.0);
    }

    #[test]
    fn world_aabb_with_translation() {
        let s = make_rect_shape("r1", 100.0, 200.0, 50.0, 30.0);
        let aabb = s.world_aabb();
        assert_eq!(aabb.min_x, 100.0);
        assert_eq!(aabb.min_y, 200.0);
        assert_eq!(aabb.max_x, 150.0);
        assert_eq!(aabb.max_y, 230.0);
    }

    #[test]
    fn is_container_for_frame() {
        let mut s = make_rect_shape("f1", 0.0, 0.0, 100.0, 100.0);
        s.shape_type = ShapeType::Frame;
        s.geometry = ShapeGeometry::Frame {
            width: 100.0,
            height: 100.0,
            clip_content: true,
        };
        assert!(s.is_container());
        assert!(s.clips_content());
    }

    #[test]
    fn is_not_container_for_rectangle() {
        let s = make_rect_shape("r1", 0.0, 0.0, 100.0, 100.0);
        assert!(!s.is_container());
        assert!(!s.clips_content());
    }

    #[test]
    fn geometry_dimensions() {
        let g = ShapeGeometry::Ellipse {
            width: 80.0,
            height: 60.0,
        };
        assert_eq!(g.dimensions(), (80.0, 60.0));
    }

    #[test]
    fn shape_type_equality() {
        assert_eq!(ShapeType::Rectangle, ShapeType::Rectangle);
        assert_ne!(ShapeType::Rectangle, ShapeType::Ellipse);
    }
}
